dbms/src/DataStreams/IProfilingBlockInputStream.h

```diff
@@ -187,7 +187,7 @@ class IProfilingBlockInputStream : public IBlockInputStream
     bool enabled_extremes = false;
 
     /// Information about the approximate total number of rows is collected in the parent source.
-    bool collected_total_rows_approx = false;
+    std::atomic_bool collected_total_rows_approx {false};
```

Comment:
Non obvious.

If this should be atomic, why we still have non atomic sequence of operations:
```
    if (collected_total_rows_approx)
        return;

    collected_total_rows_approx = true;
```

instead of test-and-set?

Can you link TSan report?

Reply:
You can split this PR, because improvements in tests are ready to merge.